src/app/vertextool/qgsvertextool.cpp

```diff
@@ -2165,10 +2165,46 @@ void QgsVertexTool::moveVertex( const QgsPointXY &mapPoint, const QgsPointLocato
 
     QgsPoint pt( layerPoint );
     if ( QgsWkbTypes::hasZ( dragLayer->wkbType() ) && !pt.is3D() )
+    {
       pt.addZValue( defaultZValue() );
+      if ( !addingAtEndpoint )
+      {
+        // try to linearly interpolate z from adjacent vertices
+        const QgsPoint pointBefore = geomTmp->vertexAt( QgsVertexId( vid.part, vid.ring, vid.vertex - 1 ) );
+        const QgsPoint pointAfter = geomTmp->vertexAt( vid );
+        // we can only do this if the adjacent vertices HAVE valid z values
+        if ( !std::isnan( pointBefore.z() ) && !std::isnan( pointAfter.z() ) )
+        {
+          const double length = pointBefore.distance( pointAfter );
+          const double distance = pointBefore.distance( pt );
+          if ( length )
+          {
+            pt.setZ( pointBefore.z() + ( pointAfter.z() - pointBefore.z() ) * distance / length );
+          }
+        }
```

Comment:
I find the interpolation method used here quite weird. I'd expect that the interpolated value would be computed based on the ratio of `pt.distance( pointBefore )` and `pt.distance( pointAfter )`, so that `pt.z()` is always between `pointBefore.z()` and `pointAfter.z()`

In the current implementation `pt.z()` can far exceed `pointAfter.z()` if `distance > length`.
Is this intentional?
(same for M values)